linker-private-library/src/main/scala/org/scalajs/linker/runtime/RuntimeLong.scala

```diff
@@ -910,185 +777,142 @@ object RuntimeLong {
-        val bDouble = asSafeDouble(blo, bhi)
-        val rDouble = aDouble / bDouble
-        hiReturn = unsignedSafeDoubleHi(rDouble)
-        unsignedSafeDoubleLo(rDouble)
-      } else {
-        // 0L, because b > a
-        hiReturn = 0
-        0
-      }
+      if (isUnsignedSafeDouble(bhi))
+        fromUnsignedSafeDouble(asSafeDouble(alo, ahi) / asSafeDouble(blo, bhi))
+      else
+        0L // 0L, because b > a
     } else {
       if (bhi == 0 && isPowerOfTwo_IKnowItsNot0(blo)) {
         val pow = log2OfPowerOfTwo(blo)
-        hiReturn = ahi >>> pow
-        (alo >>> pow) | (ahi << 1 << (31-pow))
+        pack((alo >>> pow) | (ahi << 1 << (31-pow)), ahi >>> pow)
       } else if (blo == 0 && isPowerOfTwo_IKnowItsNot0(bhi)) {
         val pow = log2OfPowerOfTwo(bhi)
-        hiReturn = 0
-        ahi >>> pow
+        pack(ahi >>> pow, 0)
       } else {
         unsignedDivModHelper(alo, ahi, blo, bhi, askQuotient = true)
       }
     }
   }
 
   @inline
-  def remainder(a: RuntimeLong, b: RuntimeLong): RuntimeLong = {
-    val lo = remainderImpl(a.lo, a.hi, b.lo, b.hi)
-    new RuntimeLong(lo, hiReturn)
-  }
+  def remainder(alo: Int, ahi: Int, blo: Int, bhi: Int): Long =
+    remainderImpl(alo, ahi, blo, bhi)
 
-  def remainderImpl(alo: Int, ahi: Int, blo: Int, bhi: Int): Int = {
+  def remainderImpl(alo: Int, ahi: Int, blo: Int, bhi: Int): Long = {
     if (bothZero(blo, bhi))
       throw new ArithmeticException("/ by zero")
 
     if (isInt32(alo, ahi)) {
       if (isInt32(blo, bhi)) {
-        val lo = alo % blo
-        hiReturn = lo >> 31
-        lo
+        fromInt(alo % blo)
       } else {
         // Either a == Int.MinValue && b == (Int.MaxValue + 1), or (abs(b) > abs(a))
-        if (alo == Int.MinValue && (blo == 0x80000000 && bhi == 0)) {
-          hiReturn = 0
-          0
-        } else {
-          // a, because abs(b) > abs(a)
-          hiReturn = ahi
-          alo
-        }
+        if (alo == Int.MinValue && (blo == 0x80000000 && bhi == 0))
+          0L
+        else
+          pack(alo, ahi) // a, because abs(b) > abs(a)
       }
     } else {
-      val aAbs = inline_abs(alo, ahi)
-      val bAbs = inline_abs(blo, bhi)
-      val absRLo = unsigned_%(aAbs.lo, aAbs.hi, bAbs.lo, bAbs.hi)
-      if (ahi < 0) inline_negate_hiReturn(absRLo, hiReturn)
-      else absRLo
+      val aAbs = abs(alo, ahi)
+      val bAbs = abs(blo, bhi)
+      // Calls back into toInt() and shr()
+      val absR = unsigned_%(aAbs.toInt, (aAbs >>> 32).toInt, bAbs.toInt, (bAbs >>> 32).toInt)
+      if (ahi < 0)
+        -absR // calls back into sub()
+      else
+        absR
     }
   }
 
   @inline
-  def remainderUnsigned(a: RuntimeLong, b: RuntimeLong): RuntimeLong = {
-    val lo = remainderUnsignedImpl(a.lo, a.hi, b.lo, b.hi)
-    new RuntimeLong(lo, hiReturn)
-  }
+  def remainderUnsigned(alo: Int, ahi: Int, blo: Int, bhi: Int): Long =
+    remainderUnsignedImpl(alo, ahi, blo, bhi)
```

Comment:
Another one :)

Reply:
This one is actually important. Same for `divideUnsigned`. I added comments.